src/goto-programs/goto_convert.cpp

```diff
@@ -2242,3 +2243,47 @@ void goto_convert(
 
   ::goto_convert(to_code(symbol.value), symbol_table, dest, message_handler);
 }
+
+void goto_convertt::generate_thread_block(
+  code_blockt &dest_block,
+  const code_blockt &thread_body)
+{
+  // pseudo code for the generated
+  // instructions:
+  //
+  // A: START_THREAD : C
+  // B: GOTO Z
+  // C: SKIP
+  // D: BLOCK   --> actual thread body.
+  // E: END_THREAD
+  // Z: SKIP
+
+  // create the necessary names for the labels,
+  // these are used to jump in and out of the thread block.
+  ++thread_started_counter;
+  const std::string &label_name_c = "__CPROVER_TS_" +
+    std::to_string(thread_started_counter);
+  ++thread_started_counter;
+  const std::string &label_name_z = "__CPROVER_TS_" +
+    std::to_string(thread_started_counter);
+
+  codet tmp_a(ID_start_thread);
+  tmp_a.set(ID_destination, label_name_c);
+
+  code_gotot tmp_b(label_name_z);
```

Comment:
In my opinion you should not be generating `code_gotot` here, but instead generate goto program instructions. That would make those artificially generated labels unnecessary.